From 95e840f418007fe0d577a73590e47b9dabefbe50 Mon Sep 17 00:00:00 2001
From: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>
Date: Thu, 13 Aug 2015 23:40:50 +0200
Subject: tfm: duct tape misscompile on armhf

since gcc-5 it seems fp_mul_comba does not produce the correct code
anymore atleast on armhf.

fp_mul_comba32(
4GCHC4UTF6B5TE86JLJ2UUTQFDPCFNUI8F89S4T2G95GSDORB9PHKGJFG14BV8GONRTE3FV26GNV4NK14EMSMIFUKTN9RCT1ANKNDGH8HC3HJA23NK3NOE6IJOLF9VI4FQF3B49J1SS9FL3BF4D04BHA0L8FFQENPLIKO78F0FC38E46DDBQLDJGBM2T4FI7SS9P0OVE1QJ9Q ,

5E35JGJA52RUHS6FN60OJNCVTNI5CHSVM6ASV3PN8APTBI8G3LHBF77RK9DQ5ED412BG0PFFNK4P2TDNBQSQNATL6HMVSBGIDBEMI4VV8OH7O3OESAD1NM1AD3MJPDJO9QRR2L4MHRBD96PR2OPG5HNAKR9A9QSG8CGO33F27INOVMRCST3MAQI5FMS6IN23KIPD1L2QPTJQJ );

returns

IBF17FL9HQU0D7IL2S9H9LDT5K9H8BGSS89UIJ5SE7 92FSLE2 8P6D1UG6NTLJSVQTBLOOSH8FN02JTHVS2HRD758Q94COJF43LFEFPLIQREDMBHQ35UK1PAGIM7J4HGGDUSJMMRG7I1PMBVFU4AJQ0L3IFJBEA197Q88CB6AFU9G5HCNOTGMOH82JNTL6SOLI64NTJ1S1A6LNETRT8L43397SS0PR2TPTS3RBAAE8BU0O2C5JEDMFLGUTBRRUE

While the correct answer would be

IBF17FL9HQU0D7IL2S9H9LDT5K9H8BGSS89UIJ5SE7 A8N0JMB 8P6D1UG6NTLJSVQTBLOOSH8FN02JTHVS2HRD758Q94COJF43LFEFPLIQREDMBHQ35UK1PAGIM7J4HGGDUSJMMRG7I1PMBVFU4AJQ0L3IFJBEA197Q88CB6AFU9G5HCNOTGMOH82JNTL6SOLI64NTJ1S1A6LNETRT8L43397SS0PR2TPTS3RBAAE8BU0O2C5JEDMFLGUTBRRUE

The -O2 -> -O1 gets it right. This duct tape patch remains here until
someone has time to investigate further.

Patch-Name: tfm-duct-tape-misscompile-on-armhf.patch
Signed-off-by: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>
---
 libclamav/tomsfastmath/mul/fp_mul_comba.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libclamav/tomsfastmath/mul/fp_mul_comba.c b/libclamav/tomsfastmath/mul/fp_mul_comba.c
index d0b6c3b..1474d6a 100644
--- a/libclamav/tomsfastmath/mul/fp_mul_comba.c
+++ b/libclamav/tomsfastmath/mul/fp_mul_comba.c
@@ -12,6 +12,8 @@
 
 */
 
+#pragma GCC optimize ("O1")
+
 #include "bignum_fast.h"
 
 #if defined(TFM_PRESCOTT) && defined(TFM_SSE2)
